cpp/src/arrow/compute/exec.cc

```diff
@@ -118,19 +125,25 @@ std::string ExecBatch::ToString() const {
 ExecBatch ExecBatch::Slice(int64_t offset, int64_t length) const {
   ExecBatch out = *this;
   for (auto& value : out.values) {
-    if (value.is_scalar()) continue;
-    value = value.array()->Slice(offset, length);
+    if (value.is_scalar()) {
+      // keep value as is
+    } else if (value.is_array()) {
+      value = value.array()->Slice(offset, length);
```

Comment:
One of the goals of `ExecBatch` was to allow for highly efficient (zero allocation) slicing.  `array()->Slice` forces a lot of shared_ptr copies and should be avoided.  I think this changes the semantics of `ExecBatch` in an undesirable way

Reply:
I may be missing something. This `Slice` invocation also appears in the pre-PR code, plus the code here is within `ExecBatch::Slice`, so I would expect the `Slice` invocation to be acceptable here.

Reply:
I don't think you're missing anything.  It's just my inability to read :laughing: For some reason I thought we implemented slicing by simply reducing the length.  Sorry for the confusion.